[IV-21-22] Objetivo 5

morevi opened the pull request:
- Closes #56 
- Closes #57 
- También se actualiza el readme en dockerhub con el mismo workflow.
- Documentación sobre la elección realizada en [docker.md](https://github.com/morevi/jobcontrol/blob/objetivo-5/docs/docker.md)

## Checklist
* [x]  ¿Se han seguido las buenas prácticas en las órdenes que construyen la imagen?
* [x]  ¿La imagen incluye sólo y exclusivamente la infraestructura necesaria para pasar los tests?
* [x]  ¿Se han documentado y enlazado los commits a las imágenes que se han testeado?
* [x]  ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los tests que se van a lanzar?
* [x]  ¿Se han establecido claramente los criterios de búsqueda y de aceptación de la imagen base?

6e29b54 se me ha colado mientras hacia pruebas 🤦🏼

morevi committed:
Choose a base image #56

Use python:3.10.1-slim-bullseye image

morevi committed:
Add Dockerfile #56

The image will contain python, and the proyect dependencies.

morevi committed:
Add tasks for building, pushing, and running #56

This helps doing these tasks locally

morevi committed:
Add workflow to build and push the image #57

Build and push to dockerhub

morevi committed:
Test commit and pushes

morevi committed:
Fix missing job key #57

morevi committed:
Add repo name to Dockerfile #56

morevi committed:
Add resolved dependencies on lock file

This avoids problems on building the image

morevi committed:
Add lock file to Dockerfile to freeze install #56

morevi committed:
Use an updated version of this action #57

morevi committed:
Sync README to Dockerhub #57

Uses an new step on the action to do it.

morevi committed:
Link documentation about base image #56

Has been tested
- [python](https://hub.docker.com/_/python)
- [debian](https://hub.docker.com/_/debian)
- [ubuntu](https://hub.docker.com/_/ubuntu)

morevi commented:
⛹ Revisores → @Antobio17 @gomares @amerigal

Antobio17 commented on .github/workflows/dockerhub.yml:
```diff
@@ -0,0 +1,38 @@
+  push:
+    branches:
+      - main
+      - objetivo-5
```
El problema de esto es que, en otro objetivo (u otra rama diferente) haces un push modificando algo relacionado con el contenedor no va a subirse a Dockerhub. Por ejemplo si esos cambios están relacionado con los tests para poder ser ejecutados al no actualizarse la imagen fallarán.

Antobio17 commented on docs/docker.md:
```diff
@@ -0,0 +1,22 @@
+## Requisitos
+
+- Menor tamaño,
+- Acceso a mayor cantidad de paquetes,
```
No se si es la mejor opción para elegir un contenedor, cuantos mas paquetes más posibilidades de encontrarte deprecations o posibles fallos a futuro.

alexespana replied:
Pienso igual que mi compañero, no creo que tener la mayor cantidad de paquetes sea un requisito razonable, ya que puedes tener gran cantidad de paquetes que realmente son innecesarios.

Antobio17 commented on docs/docker.md:
```diff
@@ -0,0 +1,22 @@
+Tanto las imagenes `Python3.10-bullseye` (_bullseye_ es debian) como `debian:bookworm` (aún en testing, no la usaremos), alcanzan la version 3.10.1 de python. Sin embargo Debian como Ubuntu, en sus versiones estables, se quedan en python 3.9.x.
+
+`python3.10-bullseye` cumple los requisitos de versiones y de acceso a paquetes. Pero `python3.10-slim-bullseye`, también cumple el requisito de reducir el tamaño.
+
```
Creo que faltaría haber probado los distintos contenedores base seleccionados de inicio y comparar por ejemplo las facilidades o inconvenientes para el montaje de la imagén, ver si hay una diferencia significativa de tiempo referente a la ejecución de los tests, etc.

Antobio17 reviewed (commented):
Aquí te dejo mi revisón :+1:

alexespana commented on .github/workflows/dockerhub.yml:
```diff
@@ -0,0 +1,38 @@
+      # only if these files had changes
+      - Dockerfile
+      - pyproject.toml
+      - .github/workflows/dockerhub.yml
```
Creo que no tiene sentido que añadas a paths el mismo fichero del workflow, realmente cuando tendrías que actualizar la imagen a Dockerhub sería cuando cambiaran las dependencias o también el readme en este caso. Por cierto, el archivo .lock sería necesario añadirlo también.

alexespana replied:
Además, si se modificase el README pero no el resto, tendrías el readme desactualizado en DockerHub, aunque es un fallo menor.

morevi replied:
Se me había quedado de ir haciendo pruebas. Lo del Readme, no me parecía correcto reconstruir la imagen cada vez que se modificase el Readme, por eso lo dejé fuera. También podría hacerse en un workflow aparte. Por ahora lo voy a dejar así, si JJ pide el cambio, no hay problema

alexespana commented on docs/docker.md:
```diff
@@ -0,0 +1,22 @@
+
+- Menor tamaño,
+- Acceso a mayor cantidad de paquetes,
+- Última version de python y de SO lts disponible.
```
No por ser la ultima versión de python significa que sea la más adecuada para tu contenedor.

morevi replied:
Pero es la 3.10.1 la que tengo instalada localmente, ¿eso la hace más adecuada?

morevi replied:
Vale, estoy viendo que usar 3.10 no es tan buena idea, acaba de salir hace escasos días

alexespana commented on docs/docker.md:
```diff
@@ -0,0 +1,22 @@
+
+Tanto las imagenes `Python3.10-bullseye` (_bullseye_ es debian) como `debian:bookworm` (aún en testing, no la usaremos), alcanzan la version 3.10.1 de python. Sin embargo Debian como Ubuntu, en sus versiones estables, se quedan en python 3.9.x.
+
+`python3.10-bullseye` cumple los requisitos de versiones y de acceso a paquetes. Pero `python3.10-slim-bullseye`, también cumple el requisito de reducir el tamaño.
```
Para mejorar la documentación, podrías incluir una tabla comparativa con los tiempos de creación y los tamaños de la imágen con distintas imágenes base, de forma que tengas algo claramenete objetivo con lo que realizar la elección.

alexespana reviewed (changes requested):
Buenas @morevi ,  creo que los errores principales están en la documentación. Échale un ojo a lo que te he comentado más abajo.

morevi committed:
Remove paths and add lock file #57

Removing the path helps executing the workflow from any branch.
Adding the lock file helps installing the new deps.

morevi committed:
Drop the python version used to 3.8 #56

This version is more stable and has a bigger hold on PyPI repos.
3.10 is just too new.

morevi committed:
Choose a different base image #56

And document the choice.

morevi commented:
@alexespana @Antobio17 gracias por la revisión, he hecho algunos cambios teniendo en cuenta lo que dijisteis.
Lo mismo ya os había llegado notificación, de usar el boton de "re request review", si es así, disculpad 😅

Antobio17 commented:
> @alexespana @Antobio17 gracias por la revisión, he hecho algunos cambios teniendo en cuenta lo que dijisteis.
> 
> Lo mismo ya os había llegado notificación, de usar el boton de "re request review", si es así, disculpad 😅 

De nada! Si quieres que le volvamos a echar un vistazo o cualquier cosa o duda que te podamos ayudar nada mas que decirlo 👍 
Siempre teniendo en cuenta que no somos JJ 😆 esto es mas bien un debate de opiniones y si crees que los cambios sugeridos son buena idea nada mas que adoptarlos o incluso mejorarlos.

alexespana commented:
> @alexespana @Antobio17 gracias por la revisión, he hecho algunos cambios teniendo en cuenta lo que dijisteis. Lo mismo ya os había llegado notificación, de usar el boton de "re request review", si es así, disculpad sweat_smile

No hay de qué! Digo lo mismo que mi compañero, cualquier cosa que necesites solo tienes que decirlo :smile:

morevi commented:
@JJ Listo para revisión

gomares commented:
Vaya, parece que llego tarde a la revisión. Estuve de viaje todo el fin de semana por un evento y no ví la mención que me hiciste @morevi . No obstante voy a echarle un ojo aunque creo que los problemas principales ya los han tratado mis compañeros. Siento no haber podido estar ahí cuando era necesario.

JJ reviewed (changes requested):
Todo bien, pero el milestone no incluye esto en su producto mínimamente viable (que tampoco está demasiado bien definido)

morevi commented:
@JJ he modificado el milestone, ¿puedo mergear ya? o Listo para revisión.

JJ commented:
> @JJ he modificado el milestone, ¿puedo mergear ya? o Listo para revisión.

No se trata de que edites el milestone, porque ahora resulta que hay un montón de issues (y el PR anterior #54) que not tienen absolutamente nada que ver con este milestone. Se trata de que *realmente* uséis los milestones para lo que son, y no simplemente para asignarles aleatoriamente un PR a cada uno. Aparte, tal como está ahora no queda claro qué es lo que hay que hacer antes ni qué funcionalidades tiene. Por favor, revisa bien el concepto de milestone y deja el que estaba como estaba y/o añádele lo necesario para que *realmente* incluya lo que se supone que has hecho dentro de él.

morevi commented:
@JJ Creo que ahora si 😓 . El milestone describe productos, y los requisitos para cumplirlo. He sacado los issues de CI a un nuevo milestone, de forma que los issues y PRs que hay describe lo que se ha hecho para cumplir los requisitos.
Listo para revisión

JJ reviewed (approved):
:+1:
